### PR TITLE
fix(knowledge): save embedding model change directly on an empty base

### PR DESCRIPTION
**Pull request by eeee0717**
> ### 🚨 Branch strategy — read before opening this PR
>
> The v2 refactor has merged into `main`, so **`main` is the default branch for active development** (v1 and v2 code currently coexist there — expect large, breaking changes).
>
> - **Active development** (features, refactors, optimizations, fixes for the current codebase) → target **`main`** (the default base).
> - **v1 maintenance** (hotfixes and subsequent v1 releases) → branch from and target **`v1`**, _not_ `main`.
>
> A v1 fix does **not** auto-carry to `main`: if the same bug exists on `main`, open a separate forward-port PR targeting `main`. Before touching subsystems being replaced, read `docs/references/data/` and watch for `@deprecated` markers — they flag code being deleted.

### What this PR does

Before this PR: changing a knowledge base's embedding model in the RAG config panel always routed through the "restore into a new knowledge base" flow — creating a brand-new base with a new id and re-embedding everything — even when the base had zero items and nothing to re-embed.

After this PR: when a knowledge base has zero items, changing the embedding model is saved in place through a normal `PATCH`. The panel now shows "Save" instead of "Rebuild" in that case. `KnowledgeBaseService.update()` enforces the same empty-base guard server-side (by querying live item count), so the change is still rejected once the base has items, regardless of what the client believes.

Fixes #

### Why we need it and why it was done in this way

The following tradeoffs were made:
- `embeddingModelId`/`dimensions` were previously excluded from the PATCH schema entirely, on the assumption changing either always invalidates existing data. This PR loosens that only within a server-enforced "base has zero items" invariant, rather than trusting the client's item count.

The following alternatives were considered:
- Keep routing through restore but relabel the button — rejected because restore always creates a new knowledge base (new id), which is the wrong semantics for editing config on a base that has no data yet.

Links to places where the discussion took place: N/A

### Breaking changes

None. This is additive: previously-disallowed PATCH fields are now allowed only in the narrow empty-base case; the restore flow for a base that already has items is unchanged.

### Special notes for your reviewer

The item-count guard lives in `KnowledgeBaseService.update()` and re-queries the DB itself rather than trusting the caller, to stay correct if an item is added concurrently while the RAG config panel is open.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Branch: This PR targets the correct branch — `main` for active development, `v1` for v1 maintenance fixes
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. Check this only when the PR introduces or changes a user-facing feature or behavior.
- [x] Self-review: I have reviewed my own code (e.g., via [`/gh-pr-review`](/.claude/skills/gh-pr-review/SKILL.md), `gh pr diff`, or GitHub UI) before requesting review from others

### Release note

```release-note
fix(knowledge): changing the embedding model on a knowledge base with no items now saves immediately instead of requiring a full rebuild.
```

### CODE REVIEW

File: src/shared/data/api/schemas/knowledges.ts [hunk 1]
```diff
@@ -44,6 +48,18 @@ export const UpdateKnowledgeBaseSchema = KnowledgeBaseEntitySchema.pick(KNOWLEDG
     documentCount: KnowledgeBaseEntitySchema.shape.documentCount,
     hybridAlpha: KnowledgeBaseEntitySchema.shape.hybridAlpha
   })
+  .superRefine((value, ctx) => {
+    // Paired like create/restore: a vector base needs both, a BM25-only base
+    // needs neither. Only enforced when the caller is actually touching one of
+    // them — omitting both leaves the existing pairing untouched.
+    if ((value.embeddingModelId !== undefined) !== (value.dimensions !== undefined)) {
```
Comment (AtomsH4) [inline src/shared/data/api/schemas/knowledges.ts]: > This review comment was translated automatically.

**[A3]** The refine here only checks if both fields are present, but does not check their null pairing. `{ embeddingModelId: null, dimensions: 1536 }` can currently pass the schema; after proceeding into the service on a completed base, `validateDimensionsForEmbeddingModel(null, 1536)` will not throw an error either, and eventually it will hit `knowledge_base_status_error_check` type DB CHECK/raw SQLite, instead of DataApi validation. Existing `KnowledgeBaseSchema` and DB both require both fields to be null for BM25-only bases. It is suggested to reject null model + non-null dimensions at the schema or service layer, and add corresponding tests.

---
<details>
<summary>Original Content</summary>

**[A3]** 这里的 refine 只检查两个字段是否都出现，但没有检查它们的 null pairing。`{ embeddingModelId: null, dimensions: 1536 }` 现在能通过 schema；在 completed base 上继续进入 service 后，`validateDimensionsForEmbeddingModel(null, 1536)` 也不会报错，最后会撞到 `knowledge_base_status_error_check` 这类 DB CHECK/raw SQLite，而不是 DataApi validation。现有 `KnowledgeBaseSchema` 和 DB 都要求 BM25-only base 两个字段都为 null，建议在 schema 或 service 层拒绝 null model + non-null dimensions，并补对应测试。
</details>

File: src/renderer/pages/knowledge/panels/ragConfig/RagConfigPanel.tsx [hunk 2]
```diff
@@ -68,21 +72,47 @@ const ActiveRagConfigPanel = ({ base, onRestoreBase }: RagConfigPanelProps) => {
...
       return
     }
 
-    if (embeddingModelChanged) {
+    if (requiresRestore) {
       onRestoreBase(base, { embeddingModelId: values.embeddingModelId })
       return
     }
 
+    if (canSaveEmbeddingModelDirectly) {
```
Comment (AtomsH4) [inline src/renderer/pages/knowledge/panels/ragConfig/RagConfigPanel.tsx:99]: > This review comment was translated automatically.

**[A1]** The direct save path here reuses the above `canSubmit = canSave || embeddingModelChanged`. In the old logic, `embeddingModelChanged` would only trigger a rebuild without submitting other form fields; now when `itemCount === 0`, it executes `save(values, ...)`, and `buildKnowledgeRagConfigPatch` will PATCH all dirty fields together. The result is that after a user clears/sets chunkSize to 0 or makes overlap >= size, as long as they then change the embedding model, the button remains clickable and falls through to toast/server-side validation. I suggest splitting the submission conditions by path: direct save should use `canSave`, and only `requiresRestore` should allow submission solely based on `embeddingModelChanged`, plus add a regression test for empty-base + invalid chunk.

---
<details>
<summary>Original Content</summary>

**[A1]** 这里的直接保存路径复用了上面的 `canSubmit = canSave || embeddingModelChanged`。旧逻辑里 `embeddingModelChanged` 只会进入 rebuild，不会提交其它表单字段；现在 `itemCount === 0` 时会执行 `save(values, ...)`，而 `buildKnowledgeRagConfigPatch` 会把所有脏字段一起 PATCH。结果是用户把 chunkSize 清空/设成 0 或让 overlap >= size 后，只要再改 embedding model，按钮仍可点击并落到 toast/服务端校验。建议把提交条件按路径拆开，例如直接保存走 `canSave`，只有 `requiresRestore` 才允许单靠 `embeddingModelChanged` 提交，并补一个 empty-base + invalid chunk 的回归测试。
</details>

**Review by AtomsH4 — COMMENTED** [review]
Reviewed the empty-knowledge-base embedding-model save path. Found two warnings: one UI submit-state regression in the new direct-save path, and one PATCH schema validation gap for embeddingModelId/dimensions pairing.

**Comment by eeee0717** [comment]
Thanks for the review — both findings confirmed and fixed in b992c5b20.

**Fixed**
- **[A1]** `canSubmit` was `canSave || embeddingModelChanged`, so an invalid chunk-size/overlap combo plus an embedding-model change on an empty base left Save enabled and submitted the whole invalid form. The restore branch could safely bypass `canSave` because it only ever reads `embeddingModelId` off the base and ignores the dirty draft — the new direct-save branch doesn't have that luxury, since it submits the full form. Now gated on `requiresRestore` instead: `canSubmit = canSave || requiresRestore`, so direct-save inherits normal chunk validation while restore keeps working exactly as before. Added a regression test for the invalid-chunk + empty-base + model-change combo, plus one confirming the restore path is still submittable despite invalid chunk fields (that assumption is now pinned by a test, not just a comment).
- **[A3]** The `UpdateKnowledgeBaseSchema` pairing refine only checked presence (`!== undefined`), so `{embeddingModelId: null, dimensions: 1536}` passed schema validation and `validateDimensionsForEmbeddingModel` (which only guards the opposite direction — model set, dimensions missing), then hit the `knowledge_base_status_error_check` DB CHECK as a raw constraint violation. Added a second refine check requiring both-null or both-non-null when both fields are present. Scoped to the schema layer only (not the shared `validateDimensionsForEmbeddingModel` helper, which `create()` also uses and deliberately tolerates a stray `dimensions` on a model-less base) — `update()` has exactly one caller and it's always schema-validated, so this fully closes the gap. Added tests for both half-null directions.

Both fixes verified with `pr-review-toolkit:code-reviewer` (no new issues) and `pr-review-toolkit:pr-test-analyzer` (flagged the missing restore-path regression test, now added). Full `tsgo` typecheck (node + web) and the knowledge-domain test suite pass.

**Review by AtomsH4 — APPROVED** [review]
LGTM. The two previous warnings are addressed: the direct-save path now respects form validation, and the PATCH schema rejects half-null embeddingModelId/dimensions pairs. Latest CI is passing.

File: src/renderer/pages/knowledge/panels/ragConfig/RagConfigPanel.tsx [hunk 3]
```diff
@@ -68,21 +72,51 @@ const ActiveRagConfigPanel = ({ base, onRestoreBase }: RagConfigPanelProps) => {
...
+        try {
+          dimensions = await fetchDimensions(values.embeddingModelId)
+        } catch (error) {
+          window.toast.error(formatErrorMessageWithPrefix(error, t('message.error.get_embedding_dimensions')))
+          return
+        }
+      }
+
+      try {
+        await save(values, { embeddingModelId: values.embeddingModelId, dimensions })
```
Comment (zhangjiadi225) [inline src/renderer/pages/knowledge/panels/ragConfig/RagConfigPanel.tsx]: > This review comment was translated automatically.

**[A1]** The direct-save path here causes empty BM25-only knowledge bases to retain `searchMode: bm25` even after adding an embedding model. The old restore flow in `KnowledgeService.restoreBase()` specifically clears `searchMode` for this "gainsEmbeddings" scenario, letting `create()` fall back to default semantic search—avoiding the case where a user has configured an embedding model but still only gets full-text search. Now `save(values, { embeddingModelId, dimensions })` only adds the model and dimensions, and `buildKnowledgeRagConfigPatch()` won't submit unchanged `searchMode`, so adding a model to an empty BM25-only base saves successfully but the search behavior doesn't match the original restore semantics. I suggest setting `searchMode` to `DEFAULT_KNOWLEDGE_SEARCH_MODE` in the direct-save branch when `initialValues.embeddingModelId === null && values.embeddingModelId !== null` (and also clean up/set the corresponding `hybridAlpha` semantics), plus adding a regression test for empty BM25 base + add embedding model.

---
<details>
<summary>Original Content</summary>

**[A1]** 这里的 direct-save 路径会让空的 BM25-only 知识库在添加 embedding model 后仍保持 `searchMode: bm25`。旧的 restore 流程在 `KnowledgeService.restoreBase()` 里专门把这种 “gainsEmbeddings” 场景的 `searchMode` 置空，让 `create()` 回落到默认的语义检索，避免用户已经配置了 embedding model 但实际仍只做全文检索。现在 `save(values, { embeddingModelId, dimensions })` 只补了模型和维度，`buildKnowledgeRagConfigPatch()` 又不会提交未改动的 `searchMode`，所以从 BM25-only 空库新增模型会保存成功但检索行为不符合原 restore 语义。建议在 `initialValues.embeddingModelId === null && values.embeddingModelId !== null` 的 direct-save 分支里同时把 `searchMode` 写成 `DEFAULT_KNOWLEDGE_SEARCH_MODE`（并清理/设置对应的 `hybridAlpha` 语义），再补一个 empty BM25 base + add embedding model 的回归测试。
</details>

**Review by zhangjiadi225 — COMMENTED** [review]
> This review was translated automatically.

Found a regression in retrieval mode when adding an embedding model to an empty BM25-only knowledge base. Suggest fixing the direct-save path before merging.

---
<details>
<summary>Original Content</summary>

发现一个空 BM25-only 知识库新增 embedding model 时的检索模式回归，建议修复 direct-save 路径后再合并。
</details>

**Comment by eeee0717** [comment]
Thanks, confirmed and fixed in 10bca2c57.

**Fixed**
- Empty BM25-only bases that gain an embedding model through the new direct-save path now preserve the old restore/create semantics: if the user did not explicitly change retrieval mode, the save submits `searchMode: DEFAULT_KNOWLEDGE_SEARCH_MODE` (`hybrid`) with compatible `hybridAlpha: null`, instead of leaving the base in `bm25`.
- Added a regression test for empty BM25-only base + add embedding model directly, covering the submitted save payload and ensuring the rebuild flow is not used.

**Verified**
- `pnpm vitest run src/renderer/pages/knowledge/panels/ragConfig/__tests__/RagConfigPanel.test.tsx`
- `pnpm format`
- `pnpm lint`
- `pnpm test`
- `pnpm build:check`

**Review by zhangjiadi225 — APPROVED** [review]
LGTM. The BM25-only empty-base direct-save regression is fixed in 10bca2c57, with regression coverage added, and latest CI is green.
